### PR TITLE
Offline mode support

### PR DESCRIPTION
**Pull request by x3rt**
This PR does the following:

- Executes the `Verified` event in offline-mode
- Executes PluginApi's `PlayerJoinedEvent` in offline-mode
- Adds the Player's UserId to the base-game `Player.PlayersUserIds` dictionary in offline-mode
- Prevents PluginApi's `PlayerLeftEvent` from getting executed within the ReferenceHubs' Start method. Not sure why that is even a thing in the first place as it only causes issues.

This fixes bugs such as base-game roles spawning with no items during offline-mode when using EXILED
and most (if not all) events not being called at all in offline-mode

### CODE REVIEW

File: EXILED/Exiled.Events/Patches/Generic/OfflineModeIds.cs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+                    new(OpCodes.Call, PropertyGetter(typeof(PlayerAuthenticationManager), nameof(PlayerAuthenticationManager.UserId))),
+                    new(OpCodes.Callvirt, Method(typeof(Dictionary<string, IGameComponent>), nameof(Dictionary<string, IGameComponent>.ContainsKey))),
+                    new(OpCodes.Brtrue_S, skipLabel),
+                    new(OpCodes.Ldsfld, Field(typeof(PluginAPI.Core.Player), nameof(PluginAPI.Core.Player.PlayersUserIds))),
+                    new(OpCodes.Ldarg_0),
+                    new(OpCodes.Call, PropertyGetter(typeof(PlayerAuthenticationManager), nameof(PlayerAuthenticationManager.UserId))),
+                    new(OpCodes.Ldarg_0),
+                    new(OpCodes.Ldfld, Field(typeof(PlayerAuthenticationManager), nameof(PlayerAuthenticationManager._hub))),
+                    new(OpCodes.Callvirt, Method(typeof(Dictionary<string, IGameComponent>), nameof(Dictionary<string, IGameComponent>.Add))),
+                    new CodeInstruction(OpCodes.Nop).WithLabels(skipLabel),
```
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/OfflineModeIds.cs:83]: Need comment to explain what this transpiller does in C#

File: EXILED/Exiled.Events/Patches/Events/Player/Verified.cs [hunk 2]
```diff
@@ -39,5 +47,41 @@ private static void Postfix(PlayerAuthenticationManager __instance)
...
+            int index = newInstructions.FindIndex(x => x.opcode == OpCodes.Callvirt && x.OperandIs(Method(typeof(CharacterClassManager), nameof(CharacterClassManager.SyncServerCmdBinding)))) + offset;
+
+            // Verified.PlayerVerified(this._hub);
+            newInstructions.InsertRange(
+                index,
+                new[]
+                {
+                    new CodeInstruction(OpCodes.Ldarg_0),
+                    new CodeInstruction(OpCodes.Ldfld, Field(typeof(NicknameSync), nameof(NicknameSync._hub))),
+                    new CodeInstruction(OpCodes.Call, Method(typeof(Verified), nameof(Verified.PlayerVerified))),
```
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Events/Player/Verified.cs:78]: Missing comment

File: EXILED/Exiled.Events/Patches/Generic/OfflineModeIds.cs [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+            int index = newInstructions.FindIndex(x => x.opcode == OpCodes.Callvirt && x.OperandIs(Method(typeof(CharacterClassManager), nameof(CharacterClassManager.SyncServerCmdBinding)))) + offset;
+
+            // EventManager.ExecuteEvent(new PlayerJoinedEvent(this._hub));
+            newInstructions.InsertRange(
+                index,
+                new[]
+                {
+                    new CodeInstruction(OpCodes.Ldarg_0),
+                    new CodeInstruction(OpCodes.Ldfld, Field(typeof(NicknameSync), nameof(NicknameSync._hub))),
+                    new CodeInstruction(OpCodes.Call, Method(typeof(OfflineModeJoin), nameof(ExecuteNwEvent))),
```
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Generic/OfflineModeIds.cs:144]: Same as above

**Review by louis1706 — CHANGES_REQUESTED** [review]
Nothing to say other than transpiler need comment

**Review by louis1706 — APPROVED** [review]

**Comment by x3rt** [comment]
No Breaking Changes btw

**Comment by louis1706** [comment]
> No Breaking Changes btw

Yes it's will be merged when an other dev boy will confirm your PR

**Review by Vladislav-CS — APPROVED** [review]

**Pull request merged**